Mathlib/GroupTheory/GroupAction/Primitive.lean

```diff
@@ -0,0 +1,410 @@
+/-
+Copyright (c) 2024 Antoine Chambert-Loir. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Antoine Chambert-Loir
+-/
+
+import Mathlib.GroupTheory.GroupAction.Blocks
+import Mathlib.GroupTheory.GroupAction.Transitive
+import Mathlib.Data.Setoid.Partition
+
+/-!
+# Primitive actions
+
+## Definitions
+
+- `IsPreprimitive G X`
+a structure that says that the action of a type `G` on a type `X`
+(defined by an instance `SMul G X`) is *preprimitive*,
+namely, it is pretransitive and the only blocks are ⊤ and subsingletons.
+(The pretransitivity assumption is essentially trivial,
+because orbits are blocks, unless the action itself is trivial.)
+
+The notion which is introduced in classical books on group theory
+is restricted to group actions.
+In fact, it may be irrelevant if the action is degenerate,
+when “trivial blocks” might not be blocks.
+Moreover, the classical notion is *primitive*,
+which assumes moreover that `X` is not empty.
+
+- `IsQuasipreprimitive G X`
+a structure that says that the action of the group `G` on the type `X` is *quasipreprimitive*,
+namely, normal subgroups of `G` which act nontrivially act pretransitively.
+
+- We prove some straightforward theorems that relate preprimitivity
+  under equivariant maps, for images and preimages.
+
+## Relation with stabilizers
+
+- `isPreprimitive_of_block_order`
+relates primitivity and the fact that the inclusion
+order on blocks containing is simple.
+
+- `maximal_stabilizer_iff_preprimitive`
+an action is preprimitive iff the stabilizers of points are maximal subgroups.
+
+## Relation with normal subgroups
+
+- `IsPreprimitive.isQuasipreprimitive`
+preprimitive actions are quasipreprimitive
+
+-/
+
+open Pointwise
+
+namespace MulAction
+
+section IsTrivialBlock
+
+variable {M α N β : Type*}
+
+section monoid
+
+variable [Monoid M] [MulAction M α] [Monoid N] [MulAction N β]
+
+@[to_additive]
+theorem IsTrivialBlock.image {φ : M → N} {f : α →ₑ[φ] β}
+    (hf : Function.Surjective f) {B : Set α} (hB : IsTrivialBlock B) :
+    IsTrivialBlock (f '' B) := by
+  cases' hB with hB hB
+  · apply Or.intro_left; apply Set.Subsingleton.image hB
+  · apply Or.intro_right; rw [hB]
+    simp only [Set.top_eq_univ, Set.image_univ, Set.range_eq_univ, hf]
+
+@[to_additive]
+theorem IsTrivialBlock.preimage {φ : M → N} {f : α →ₑ[φ] β}
+    (hf : Function.Injective f) {B : Set β} (hB : IsTrivialBlock B) :
+    IsTrivialBlock (f ⁻¹' B) := by
+  cases' hB with hB hB
+  · apply Or.intro_left; exact Set.Subsingleton.preimage hB hf
+  · apply Or.intro_right; simp only [hB, Set.top_eq_univ]; apply Set.preimage_univ
```

Comment:
This has nothing to do with primitive actions, right? Can this move to an earlier file?

Reply:
moved to blocks